core/src/test/java/org/elasticsearch/action/support/replication/ReplicationOperationTests.java

```diff
@@ -437,6 +448,19 @@ public void failShard(ShardRouting replica, long primaryTerm, String message, Ex
                 fail("replica [" + replica + "] was failed");
             }
         }
+
+        @Override
+        public void markShardCopyAsStale(ShardId shardId, String allocationId, long primaryTerm, Runnable onSuccess,
+                                         Consumer<Exception> onPrimaryDemoted, Consumer<Exception> onIgnoredFailure) {
```

Comment:
Do we have a unit test that uses the `onPrimaryDemoted` consumer to test when the primary is demoted?